docs/source/schemas/dynamic_columns.rst

```diff
@@ -0,0 +1,140 @@
+4. "Find all of Tom's largest todos" is somewhat more difficult, as reverse range scans aren't supported by the
+   dynamic columns API. If we were more interested in largest todos as opposed to smallest, we could set the
+   ``ValueByteOrder`` of the ``taskSize`` column component to decreasing.
+5. "Find all of Tom's todos which cost less than 5" will not be efficient; we need to retrieve the entire row to do
+   this, because there is no sorting by sub-units.
+6. "Find all of Tom's todos which have size between 3 and 7, and cost between 5 and 10". This is achievable, though in
+   addition to the values we want to see, we also need to scan through all values with size 4-6, and those with size
+   3 and cost >10, or 7 and cost <5.
+7. "Find all of John's and Jeremy's smallest todos" is easy for a similar reason to 1. We provide an API that allows
+   for this to be done seamlessly as well.
+8. "Find the highest priority todos for everyone whose name begins with J" is NOT easy. While this
+   conceivably can be split into two parts (a prefix scan for J, and then the aforementioned getRowsColumnRange),
+   the range scan for J is potentially costly, as it needs to iterate through every todo for people whose names
+   do indeed begin with J.
+
+Using Dynamic Columns
+---------------------
+
+Storage
+=======
+
+One can use the standard ``put`` method. This can be used to add one or more columns to a row.
+
+.. code:: java
+
+    TodoTable todoTable = TodoSchemaTableFactory.of().getTodoTable(tx);
+
+    todoTable.put(
+            TodoTable.TodoRow.of(person),
+            TodoTable.TodoColumnValue.of(
+                    TodoTable.TodoColumn.of(size1, cost1), description1),
+                    TodoTable.TodoColumn.of(size2, cost2), description2));
+
+Note that if the table already consists of a value for a given row-key and dynamic column values, then this put
+will overwrite the existing value.
+
+Retrieval
+=========
+
+One can use the standard ``getRowsMultimap`` method to retrieve a multimap of rows to column values. However, note
+that this method will call the underlying transaction's ``getRows`` method, which eagerly loads the entire row into
+memory!
+
+Alternatively, one can use ``getRowsColumnRange``. This takes a collection of rows and key ranges, and returns a map of
+row-keys to BatchingVisitables of column values. These may in turn be traversed using the BatchingVisitable
+API. For example, the code below is an implementation of query 1 for an arbitrary String ``person``:
+
+.. code:: java
+
+    TodoTable todoTable = TodoSchemaTableFactory.of().getTodoTable(tx);
+    Map<TodoTable.TodoRow, BatchingVisitable<TodoTable.TodoColumnValue>> results =
+            todoTable.getRowsColumnRange(
+                    ImmutableList.of(TodoTable.TodoRow.of(person)),
+                    BatchColumnRangeSelection.create(PtBytes.EMPTY_BYTE_ARRAY, PtBytes.EMPTY_BYTE_ARRAY, 1));
```

Comment:
side note that we should really add a better API for creating an all-inclusive range

Reply:
Yep - I've filed https://github.com/palantir/atlasdb/issues/2391. This one only applies to column selections, which I guess haven't been prioritised as highly since dynamic columns aren't very widely used anyway...